### PR TITLE
st7789 lcd driver with spi

### PR DESCRIPTION
**Pull request by anjana348**
Added st7789 lcd driver
App ameba_spi_lcd for demo pattern
SPI code changes for selecting cmd or data

### CODE REVIEW

File: os/drivers/lcd/lcd_dev.c [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+  priv = filep->f_inode->i_private;
+
+  switch (cmd)
+    {
+    case LCDDEVIO_GETRUN:
+      {
+        FAR struct lcddev_run_s *lcd_run =
+            (FAR struct lcddev_run_s *)arg;
+
+        ret = priv->planeinfo[lcd_run->planeno].getrun(priv->lcd_ptr, lcd_run->row,
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/lcd_dev.c:181]: each of case should have below.
```
if (!priv->any_ops) {
    return -ENOSYS;
}
```

File: os/drivers/lcd/lcd_dev.c [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+    default:
+      {
+        if (priv->lcd_ptr->ioctl)
+          {
+            ret = priv->lcd_ptr->ioctl(priv->lcd_ptr, cmd, arg);
+          }
+        else
+          {
+            printf("ERROR: Unsupported IOCTL command: %d\n", cmd);
+            ret = -ENOTTY;
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/lcd_dev.c:330]: -ENOSYS

File: os/drivers/lcd/lcd_dev.c [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+  //pthread_mutex_init(&priv->lock, NULL);
+
+  priv->lcd_ptr = board_lcd_getdev(devno);
+  if (!priv->lcd_ptr)
+    {
+      ret = -ENODEV;
+      goto err;
+    }
+
+  ret = priv->lcd_ptr->getplaneinfo(priv->lcd_ptr, 0, &priv->planeinfo); //plane no is taken 0 here
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/lcd_dev.c:382]: then planeinfo need to be checked here.

File: os/drivers/lcd/lcd_dev.c [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+ * Input Parameters:
+ *   devno - The LCD device number.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success.  Otherwise a negated errno value is
+ *   returned to indicate the nature of the failure.
+ *
+ ****************************************************************************/
+
+int lcddev_register(int devno)
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/lcd_dev.c:358]:
```
lcd_dev_s *st7789_lcdinitialize(struct spi_dev_s *spi)
{
	FAR struct st7789_dev_s *dev;
	do_something;
	return dev;
}

int lcddev_register(FAR struct lcd_dev_s *dev)
{
	FAR struct lcddev_dev_s *priv;
	...
	priv->lcd_ptr = dev;
	...
	register_driver(path, g_lcddev_fops, 066, priv); 
}
```

File: os/drivers/lcd/st7789.c [hunk 5]
```diff
@@ -0,0 +1,975 @@
...
+ *   bits  - Number of SPI bits
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+static void st7789_select(FAR struct spi_dev_s *spi)
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/st7789.c:308]: I think selelct & unselect is unnecessary function, only 1line

File: os/drivers/lcd/lcd_dev.c [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+  priv = (struct lcddev_dev_s *)kmm_zalloc(sizeof(struct lcddev_dev_s));
+
+  if (!priv)
+    {
+      return -ENOMEM;
+    }
+
+  //pthread_mutex_init(&priv->lock, NULL);
+
+  priv->lcd_ptr = board_lcd_getdev(devno);
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/lcd_dev.c:375]: so it need to be removed here. it should be hided from board specific code. because lcd_dev is generic virtual driver

File: os/board/rtl8730e/src/rtl8730e_st7789_lcd.c [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    	}
+  	else
+    	{
+      		lldbg("SPI1 port bound to LCD %d\n", devno);
+      		return g_lcd;
+    	}
+  	return NULL;
+}
+
+void board_lcd_initialize(void)
```
Comment (Taejun-Kwon) [inline os/board/rtl8730e/src/rtl8730e_st7789_lcd.c:73]: 1. please refer to how another driver initialized. you can refer to artik055_alc5658.c & audio.c
2. rename to rtl8730e_st7789.c
3. rename oo rtl8730_st7789_initialize this need to be called once in rtl8730e_boot.c
4. so above board_lcd_getdev should be removed.

option 2, 3 is acceptable if it should handle some specific operation like a gpio interrupt.
otherwise you can call as a below in rtl8730e_boot.c
```
#if defined(CONFIG_LCD) && defined(CONFIG_LCD_ST7789) //CONFIG_LCD is necessary??
	struct spi_dev_s *spi;
	struct lcd_dev_s *dev;
	spi = up_spiinitialize(x);
	dev = st7789_lcdinitialize(spi);
	lcddev_register(dev);
#endif
```
code line 75 to 103 seems that option 2, 3 is necessary, IMO.

File: os/drivers/lcd/st7789.c [hunk 8]
```diff
@@ -0,0 +1,975 @@
...
+/****************************************************************************
+ * Name: st7789_sendcmd
+ *
+ * Description:
+ *   Send a command to the driver.
+ *
+ ****************************************************************************/
+static inline void st7789_sendcmd(FAR struct st7789_dev_s *dev, uint8_t cmd)
+{
+#ifdef CONFIG_LCD_ST7789_3WIRE
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/st7789.c:348]: this config should be defined in drivers/lcd/KConfig

**Comment by sunghan-chang** [comment]
Please find our commit rules. https://github.com/Samsung/TizenRT/wiki/Documentations#commit-rules

File: os/include/tinyara/fs/ioctl.h [hunk 9]
```diff
@@ -389,6 +389,16 @@
 #define _SPIIOCVALID(c)   (_IOC_TYPE(c) == _SPIBASE)
 #define _SPIIOC(nr)       _IOC(_SPIBASE, nr)
 
+
+#define _LCDIOCBASE     (0x1100) /* LCD character driver ioctl commands */
```
Comment (Taejun-Kwon) [inline os/include/tinyara/fs/ioctl.h:393]: _SLCDIOCBASE already defined as 0x1100 already.
Hence you should put new base value to line 104

File: apps/examples/board_specific_demo/ameba_spi_lcd_demo/Kconfig [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#
+# For a description of the syntax of this configuration file,
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config EXAMPLES_AMEBA_SPI_LCD
+	bool "AMEBA_SPI_LCD example"
+	default n
+	select AMEBASMART_SPI_LCD
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_spi_lcd_demo/Kconfig:9]: I can't find the config `AMEBASMART_SPI_LCD`. Could you check this?

File: apps/examples/board_specific_demo/ameba_spi_lcd_demo/Make.defs [hunk 11]
```diff
@@ -0,0 +1,59 @@
+###########################################################################
+#
+# Copyright 2023 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_spi_lcd_demo/Make.defs:3]: 2024

File: apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c [hunk 12]
```diff
@@ -0,0 +1,152 @@
+#include <tinyara/config.h>
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c:1]: Need to add Samsung license

File: apps/examples/board_specific_demo/ameba_spi_lcd_demo/Kconfig [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+	---help---
+		Enable AMEBA_SPI_LCD Example
```
Comment (jylee9613) [inline apps/examples/board_specific_demo/ameba_spi_lcd_demo/Kconfig:11]: What is the usage of AMEBASMART_SPI_LCD config? There is no code for AMEBASMART_SPI_LCD.
If you don't have any codes for AMEBASMART_SPI_LCD, please remove this config.

File: apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+#include <mqueue.h>
+#include <stdio.h>
+#include <stdlib.h>
+#  define ST7789_XRES       240
+#  define ST7789_YRES       320
+
+#define LCD_DEV_PATH "/dev/lcd%d"
+static pid_t g_task = 0;
+
+void test_put_area()
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c:23]: Because of the coding rules, please use `(void)` instead of `()`.

File: apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+		test_put_area();
+		count++;
+		sleep(2);
+	}
+}
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int ameba_spi_lcd_main(int argc, char *argv[])
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c:146]: Is there anything related with amebasmart? If not, let's locate this under example, not in board_specific_demo.

File: apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+#endif
+{	
+	printf("=== LCD demo ===");
+	char *args[5] = {NULL, NULL, NULL, NULL, NULL};
+	g_task = task_create("LCD SPI Demo", 100, 4096, SPIDemoShow_task, args);
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c:151]: There is no reason to make a new task.
I mean you launch the app as below:
1. kernel creates the app `ameba_spi_lcd`
2. The `ameba_spi_lcd` task creates the app `LCD SPI Demo`
3. The `LCD SPI Demo` task runs real sample code.

2nd is redundant. Let's call SPIDemoShow_task here directly.
If the default stack size is not enough, then you can set it in Makefile

https://github.com/Samsung/TizenRT/pull/6074/files#diff-67b7fda12ca22aa3f6fdc12ccbddcdc076c9ead3fc3a7d00e1bc1bfc536d92eeR60
```
APPNAME = ameba_spi_lcd
FUNCNAME = $(APPNAME)_main
THREADEXEC = TASH_EXECMD_ASYNC
PRIORITY = 4096
```

File: os/arch/arm/src/amebasmart/amebasmart_spi.c [hunk 17]
```diff
@@ -857,7 +870,7 @@ static uint16_t amebasmart_spi_send(FAR struct spi_dev_s *dev, uint16_t wd)
 	DEBUGASSERT(priv);
 
 	if (priv->role == AMEBASMART_SPI_MASTER) {
-
+
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebasmart/amebasmart_spi.c:873]: Let's revert unnecessary changes.

File: os/arch/arm/src/amebasmart/amebasmart_spi.c [hunk 18]
```diff
@@ -696,6 +696,19 @@ uint8_t amebasmart_spi1status(FAR struct spi_dev_s *dev, enum spi_dev_e devid)
  ************************************************************************************/
 int amebasmart_spi1cmddata(FAR struct spi_dev_s *dev, uint32_t devid, bool cmd)
 {
+	
+	FAR struct amebasmart_spidev_s *priv = (FAR struct amebasmart_spidev_s *)dev;
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebasmart/amebasmart_spi.c:700]: Let's use the same indentation this file already uses. It looks a tab.

File: os/arch/arm/src/amebasmart/amebasmart_spi.c [hunk 19]
```diff
@@ -696,6 +696,19 @@ uint8_t amebasmart_spi1status(FAR struct spi_dev_s *dev, enum spi_dev_e devid)
...
+        if(cmd == 0)
+		{
+				gpio_write(&priv->gpio_rs, 0);
+				DelayMs(10);
+		}
+        else
+		{
+				gpio_write(&priv->gpio_rs, 1);
+				DelayMs(10);
+		}
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebasmart/amebasmart_spi.c:711]: if (cmd == 0) {
    ....
} else {
    ....
}

File: os/arch/arm/src/amebasmart/amebasmart_spi.c [hunk 20]
```diff
@@ -696,6 +696,19 @@ uint8_t amebasmart_spi1status(FAR struct spi_dev_s *dev, enum spi_dev_e devid)
  ************************************************************************************/
 int amebasmart_spi1cmddata(FAR struct spi_dev_s *dev, uint32_t devid, bool cmd)
 {
+	
+	FAR struct amebasmart_spidev_s *priv = (FAR struct amebasmart_spidev_s *)dev;
+
+        if(cmd == 0)
+		{
+				gpio_write(&priv->gpio_rs, 0);
+				DelayMs(10);
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebasmart/amebasmart_spi.c:705]: 1. Let's use usleep or sleep.
2. Do we need this delay time?

File: apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+	for(i=100;i<=200;i+=2)
+	{
+		spi_data[i] = 0x001F & 0X00FF;
+  		spi_data[i+1] = (0x001F & 0xFF00)>>8; 
+	}
+	for(i=200;i<=400;i+=2)
+	{
+		spi_data[i] = 0xF800 & 0X00FF;
+  		spi_data[i+1] = (0xF800 & 0xFF00)>>8; 
+	}
```
Comment (kishore-sn) [inline apps/examples/board_specific_demo/ameba_spi_lcd_demo/example_spi_lcd.c:56]: There are many hard coded values above. The meaning of these values is not clear. 
Please add preprocessor definitions or comments to indicate the meaning of each value.

File: os/arch/arm/src/amebasmart/amebasmart_spi.c [hunk 22]
```diff
@@ -105,6 +105,7 @@ struct amebasmart_spidev_s {
 	uint8_t mode;               /* Mode 0,1,2,3 */
 	int role;
 	gpio_t gpio_cs0;
+	gpio_t gpio_rs;
```
Comment (abhishek-samsung) [inline os/arch/arm/src/amebasmart/amebasmart_spi.c:108]: Could you please let us know why rs is added to spidev structure. Generally spi only has mosi, miso, sclk and cs. 

Screenshot (2)

And more over, rs seems to be lcd specific. So, could you please check if its better to move the logic related to rs out of amebasmart_spi.c?

Comment (anjana348) [reply os/arch/arm/src/amebasmart/amebasmart_spi.c:108]: This pin is required to identify the data type needed by LCD. We are using below function every time to switch it from 0 to 1 based on data type(Command/data) before sending actual data from driver.
int amebasmart_spi1cmddata(FAR struct spi_dev_s *dev, uint32_t devid, bool cmd) {
	FAR struct amebasmart_spidev_s *priv = (FAR struct amebasmart_spidev_s *)dev;
	if (cmd == 0) {
		gpio_write(&priv->gpio_rs, 0);
	} else {
		gpio_write(&priv->gpio_rs, 1);
	}
	return 0;
}
Where should this function be present if not here?

Comment (Taejun-Kwon) [reply os/arch/arm/src/amebasmart/amebasmart_spi.c:108]: so you mean it is send command to spi, right? then you can use cs0.
```
#st7789.h
struct st7789_config_s {
	gpio_t gpio_cs;
	gpio_t gpio_reset;
	int (*reset)(gpio_t gpio);
};

#rtl8730csm_st7789.c

struct st7789_config_s g_rtl8730_config_dev_s {
	.gpio_cs.pin = PA_10,
	.gpio_reset.pin = PB_11,
	.reset = rtl8730_st7789_gpio_reset
};

static int rtl8730_st7789_gpio_reset(gpio_t gpio)
{
	gpio_write(gpio, GPIO_SET);
	DelayMs(10);
	gpio_write(gpio, GPIO_UNSET);
	DelayMs(10);
	gpio_write(gpio, GPIO_SET);
}

void rtl8730_st7789_initialize(void)
{
    /* Please check pinmux setting necessary or not???? */
	FAR struct spi_dev_s *spi = up_spiinitialize(LCD_SPI_PORT);
	if (!spi) {
		lldbg("spi init failed\n");
		return;
	}
	struct lcd_dev_s *dev = st7789_lcdinitialize(spi, g_rtl8730_config_dev_s);
	if (lcddev_register(dev) < 0) {
		lldbg("LCD driver register fail\n");
	} else {
		lldbg("LCD driver register success\n");
	}
}

#st7789.c
struct st7789_dev_s {
	struct lcd_dev_s dev;
	st7789_config_s *config;
	/* Private LCD-specific information follows */

	FAR struct spi_dev_s *spi;  /* SPI device */
	uint8_t bpp;                /* Selected color depth */
	uint8_t power;              /* Current power setting */
	uint16_t runbuffer[ST7789_LUT_SIZE];
};


FAR struct lcd_dev_s *st7789_lcdinitialize(FAR struct spi_dev_s *spi, st7789_config_s *config)
{
	FAR struct st7789_dev_s *priv = &g_lcddev;
	/* Initialize the driver data structure */
	priv->dev.getvideoinfo = st7789_getvideoinfo;
	priv->dev.getplaneinfo = st7789_getplaneinfo;
	priv->dev.getpower     = st7789_getpower;
	priv->dev.setpower     = st7789_setpower;
	priv->dev.getcontrast  = st7789_getcontrast;
	priv->dev.setcontrast  = st7789_setcontrast;
	priv->dev.init         = st7789_init;
	priv->dev.clear        = st7789_clear;
	priv->spi              = spi;
	priv->config = config;
	priv->spi.gpio_cs0	= config->gpio_cs;
	SPI_SETMODE(spi, CONFIG_LCD_ST7789_SPIMODE);
	SPI_SETBITS(spi, LCD_ST7789_SPI_BITS);
	SPI_SETFREQUENCY(spi, CONFIG_LCD_ST7789_FREQUENCY);

	st7789_init_cmd(priv);	
	st7789_sendcmd(priv, ST7789_DISPOFF);
	st7789_bpp(priv, ST7789_BPP);
	st7789_sendcmd(priv, ST7789_DISPON);
	st7789_fill_frame(priv, 0xFFFF, 0, 0, ST7789_XRES-1, ST7789_YRES-1);
	/* For reset */
	priv->config->reset(priv->config->gpio_reset);
	
	return &priv->dev;
}
```
@abhishek-samsung please correct this if anything is wrong

Comment (anjana348) [reply os/arch/arm/src/amebasmart/amebasmart_spi.c:108]: Can we use gpio_t in driver side code directly? I couldn't find any example for this.
#st7789.h
struct st7789_config_s {
	gpio_t gpio_cs;
	gpio_t gpio_reset;
	int (*reset)(gpio_t gpio);
};

File: os/drivers/lcd/st7789.c [hunk 23]
```diff
@@ -0,0 +1,882 @@
...
+	/* Add data prefix (9th bit shoudl be 1 ) */
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, true);
+	for (int j = 0; j < size; j += 2)
+	{
+	/* Take care of correct byte order. */
+
+		rowbuff[j] = LCD_ST7789_DATA_PREFIX | (uint16_t)txbuf[j + 1 + (i * size)];
+		rowbuff[j + 1] = LCD_ST7789_DATA_PREFIX | (uint16_t)txbuf[j + (i * size)];
+	}
+    SPI_SNDBLOCK(dev->spi, rowbuff, size);
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/st7789.c:356]: tab

**Review by Taejun-Kwon — COMMENTED** [review]
I think there's chance that both of lcd & dsp(or some other h/w) use same SPI bus.
Hence I think each of operation should call SPI_LOCK.
AFter LOCK, it should call SPI_SETMODE, SPI_SETBITS, SPI_SETFREQUENCY .
You can refer to mtd_jedec.c and jedec_lock / jedec_unlock.
That is common flash driver based following jedec interface based on SPI

File: os/board/rtl8730e/src/rtl8730e_st7789.c [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+	return OK;
+}
+
+static int rtl8730_st7789_gpio_backlight(uint8_t level) //on - 1, off - 0
+{
+	GPIO_WriteBit(GPIO_PIN_BACKLIGHT, level);
+	return OK;
+}
+
+struct st7789_config_s g_rtl8730_config_dev_s = {
```
Comment (Taejun-Kwon) [inline os/board/rtl8730e/src/rtl8730e_st7789.c:57]: move this struct to top area. we usually use below style.
```
/****************************************************************************
 * Included Files
 ****************************************************************************/
#include xxx.h

/****************************************************************************
 * Private Definitions
 ****************************************************************************/
#define SOMETHING_VALUE
..
int g_value;
/****************************************************************************
 * Private Types
 ****************************************************************************/
struct st7789_config_s g_rtl8730_config_dev_s = {
...
};

/****************************************************************************
 * Private Function Prototypes
 ****************************************************************************/
static int func_a();

```

File: os/board/rtl8730e/src/rtl8730e_st7789.c [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+void rtl8730_st7789_initialize(void)
+{
+	gpio_pins_init();
+
+	FAR struct spi_dev_s *spi = up_spiinitialize(LCD_SPI_PORT);
+	struct lcd_dev_s *dev = st7789_lcdinitialize(spi, &g_rtl8730_config_dev_s);
+	if(lcddev_register(dev) < 0) {
+		lldbg("LCD driver register fail\n");
```
Comment (Taejun-Kwon) [inline os/board/rtl8730e/src/rtl8730e_st7789.c:97]: I think you can do this later, 
Let's add LCD Debug.
You can refer to os/include/debug.h, and CONFIG_DEBUG_AUDIO_ERROR
```
#ifdef CONFIG_DEBUG_AUDIO_ERROR
#define auddbg(format, ...)    dbg(format, ##__VA_ARGS__)
#define audlldbg(format, ...)  lldbg(format, ##__VA_ARGS__)
#else
#define auddbg(...)
#define audlldbg(...)
#endif
...
```

File: os/drivers/lcd/lcd_dev.c [hunk 26]
```diff
@@ -0,0 +1,395 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (Taejun-Kwon) [reply os/drivers/lcd/lcd_dev.c:1]: Simply, you can use .tizenrt/toos/cformatter.sh
But above tool can't distingush between pointer and multiplication.
so It will change code like below. so it should be updated manually.
int *a -> int * a

**Review by Taejun-Kwon — APPROVED** [review]

**Comment by Taejun-Kwon** [comment]
please check conflicts.
also lcddbg is not defined yet.

File: os/board/rtl8730e/src/rtl8730e_boot.c [hunk 27]
```diff
@@ -376,6 +376,13 @@ void board_initialize(void)
 	board_i2c_initialize();
 	board_spi_initialize();
 	board_i2s_initialize();
+
+#ifdef CONFIG_LCD
+#ifdef CONFIG_LCD_ST7789
```
Comment (edwakuwaku) [inline os/board/rtl8730e/src/rtl8730e_boot.c:381]:
```
#if defined(CONFIG_LCD) && defined(CONFIG_LCD_ST7789)
rtl8730_st7789_initialize();
#endif
```

**Review by ewoodev — COMMENTED** [review]
https://github.com/Samsung/TizenRT/wiki/Documentations#commit-rules

Please check commit rule

**Review by ewoodev — COMMENTED** [review]
Please check that each commit contains only one topic.

For modifications to a previous commit, please rebase that commit, not the next commit.

File: os/include/tinyara/lcd/lcd.h [hunk 28]
```diff
@@ -132,61 +202,32 @@ struct lcd_planeinfo_s {
...
 
-#ifdef CONFIG_LCD_DMA_SUPPORT
-	int (*putdma)(fb_coord_t row, fb_coord_t col, fb_coord_t row2,
-				fb_coord_t col2, FAR const uint8_t *buffer);
-#endif
+	/* This is the LCD interface corresponding to which this color plane
+	 * belongs.
+	 */
+
+	FAR struct lcd_dev_s *dev;
```
Comment (Taejun-Kwon) [inline os/include/tinyara/lcd/lcd.h:209]: I think it should not be added.
```
int (*getplaneinfo)(FAR struct lcd_dev_s *dev, unsigned int planeno, FAR struct lcd_planeinfo_s *pinfo);
```
dev is already given as a first parameter

File: os/include/tinyara/lcd/lcd.h [hunk 29]
```diff
@@ -88,39 +102,95 @@ struct lcd_planeinfo_s {
...
+	 *              to allow the LCD driver to calculate the offset for partial
+	 *              writes when the buffer needs to be splited for row-by-row
+	 *              writing.
+	 *
+	 * NOTE: this operation may not be supported by the device, in which case
+	 * the callback pointer will be NULL. In that case, putrun() should be
+	 * used.
+	 */
+
+	int (*putarea)(FAR struct lcd_dev_s *dev, fb_coord_t row_start, fb_coord_t row_end, fb_coord_t col_start, fb_coord_t col_end, FAR const uint8_t *buffer, fb_coord_t stride);
```
Comment (Taejun-Kwon) [inline os/include/tinyara/lcd/lcd.h:134]: I think putarea is almost same as putrun. any special reason why you add this?

File: os/include/tinyara/lcd/lcd.h [hunk 30]
```diff
@@ -196,13 +237,20 @@ struct lcd_dev_s {
...
-};
 
-struct dsi_ops_s {
-	uint8_t (*config)(void);
-	void (*enable)(void);
-	void (*disable)(void);
-	void (*refresh)(void);
+	/* Passthrough unknown ioctl commands. */
+
+	int (*ioctl)(FAR struct lcd_dev_s *dev, int cmd, unsigned long arg);
```
Comment (Taejun-Kwon) [inline os/include/tinyara/lcd/lcd.h:243]: All LCD Driver should implement all these apis.
But I think unnecessary.

File: os/drivers/lcd/st7789.c [hunk 31]
```diff
@@ -0,0 +1,920 @@
...
+	pinfo->bpp = priv->bpp;		/* Bits-per-pixel */
+	pinfo->dev = dev;			/* The lcd device */
+	return OK;
+}
+
+/****************************************************************************
+ * Name:  st7789_getpower
+ ****************************************************************************/
+
+static int st7789_getpower(FAR struct lcd_dev_s *dev, int color)
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/st7789.c:754]: please let me know why color has added? any reason?

File: os/drivers/lcd/st7789.c [hunk 32]
```diff
@@ -0,0 +1,920 @@
...
+
+/****************************************************************************
+ * Name:  st7789_clear
+ *
+ * Description:
+ *   clear the lcd/ fill with black color
+ *
+ ****************************************************************************/
+
+static int st7789_clear(FAR struct lcd_dev_s *dev)
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/st7789.c:808]: I think this is unnecessary. user can make block screen with 'putrun', If I am wrong, please let me know.

File: os/drivers/lcd/st7789.c [hunk 33]
```diff
@@ -0,0 +1,920 @@
...
+	FAR struct st7789_dev_s *priv = &g_lcddev;
+	/* Initialize the driver data structure */
+	priv->dev.getvideoinfo = st7789_getvideoinfo;
+	priv->dev.getplaneinfo = st7789_getplaneinfo;
+	priv->dev.getpower = st7789_getpower;
+	priv->dev.setpower = st7789_setpower;
+	priv->dev.getcontrast = st7789_getcontrast;
+	priv->dev.setcontrast = st7789_setcontrast;
+	priv->dev.setorientation = st7789_setorientation;
+	priv->dev.init = st7789_init;
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/st7789.c:905]: it seems that init called only when LCD initialize. Then we ca remove this from lcd ops, and just keep this private function

File: os/include/tinyara/lcd/lcd.h [hunk 34]
```diff
@@ -88,39 +102,95 @@ struct lcd_planeinfo_s {
...
+	 * NOTE: In case of non e-ink dispalys redrawing is cheap and can be done
+	 * after each memory modification. Redrawing e-ink display is time and
+	 * energy consuming.
+	 * In order to avoid such operation (time and energy consumption) we can
+	 * implement callback function putrun without redrawing the screen.
+	 * Function putrun is called many times unless the function putarea is
+	 * implemented.
+	 */
+
+	int (*redraw)(FAR struct lcd_dev_s *dev);
```
Comment (Taejun-Kwon) [inline os/include/tinyara/lcd/lcd.h:180]: It should be removed.

File: os/include/tinyara/lcd/lcd.h (hunk 29 above)
Comment (anjana348) [reply os/include/tinyara/lcd/lcd.h:134]: putarea is for drawing a rectangle (x1,x2,y1,y2)
putrun just draws a line at a point(x,y) 
If user calls putrun multiple times instead of putarea, he needs to do some calculations ? If it is so, I will remove it

File: os/drivers/lcd/st7789.c (hunk 33 above)
Comment (anjana348) [reply os/drivers/lcd/st7789.c:905]: st7789_init - This can be called as ioctl(INIT) by user if he wants to initialize lcd in between. Should we remove this ioctl as well?

Comment (Taejun-Kwon) [reply os/drivers/lcd/st7789.c:905]: I think LCD will be initialized for every boot, and it will be kept as initialized states.
only backlight can be toggled, I mean turn on/off
How do you think?

Comment (anjana348) [reply os/drivers/lcd/st7789.c:905]: We have not provided any ioctl for backlight on/off. It is just sets on when lcd intialize . Should an ioctl be provided for it?

Comment (Taejun-Kwon) [reply os/drivers/lcd/st7789.c:905]: My domain knowledge about lcd & backlight is not good.
But as per my experience, some dimming algorithm need to be handled, so below is necessary.
1. On/Off include time value (ex. turn backlight off after x sec)
2. Dimming level(for Local dimming)

Let's consider more and apply it later, not this PR.

Comment (anjana348) [reply os/drivers/lcd/st7789.c:905]: ok, so lcd initialization (reset lcd and send initialization commands) is done at boot time. Will there be any need for initialization in between also, means from any app? If not, we will remove init ioctl as well.

Comment (Taejun-Kwon) [reply os/drivers/lcd/st7789.c:887]: @edwakuwaku Hello Edward, I have a one question.
When PM change device state to normal mode to power save mode, is each of device driver should initialized again?
You can see st7789_init in this PR, actually it initializes something through the SPI.
And st7789 will keep some state, but when pm change mode to sleep, and device lost all state because of very low voltage level, then PM should enable all device driver again.
How do you think?

Comment (edwakuwaku) [reply os/drivers/lcd/st7789.c:887]: Hello TaeJun님, as per my understanding, before going power save mode (ie. cpu go to sleep), we will save up the status of the whole system, so when the core wakeup from sleep, we do not need to reinitialize the drivers again. Previously we did some demo on UART, it is working fine (ie. no reinitialization needed, data transfer can still be performed) after wakeup from low power mode~
So, I think it should be the same for SPI, once the development is stable, we will trigger PM to see how to integrate both together
In my opinion, PM is highly dependent to application scenario, and LCDC with SPI interface seems included in the second phase for PM, we can discuss about this further

Comment (Taejun-Kwon) [reply os/drivers/lcd/st7789.c:887]: @edwakuwaku Thanks for your answer.
@anjana348 I think we can keep init in this time, and if there's no issue with PM later, then remove it at that time.

File: os/drivers/lcd/lcd_dev.c [hunk 35]
```diff
@@ -0,0 +1,388 @@
...
+	/* Allocate a new lcd_dev driver instance */
+
+	priv = (struct lcd_s *)kmm_zalloc(sizeof(struct lcd_s));
+
+	if (!priv) {
+		return -ENOMEM;
+	}
+
+	priv->dev = dev;
+	if (!priv->dev) {
```
Comment (Taejun-Kwon) [inline os/drivers/lcd/lcd_dev.c:370]: let's move this logic to line 362,
```
if (!dev) {
    return -EINVAL;
}
priv->dev = dev;
```

**Review by Taejun-Kwon — COMMENTED** [review]
please check lcddbg and debug.h

**Review by ewoodev — COMMENTED** [review]
Could you please refer master branch commits?

Please do not add commit about new review.

File: apps/examples/lcd_test/Kconfig [hunk 36]
```diff
@@ -0,0 +1,14 @@
+#
```
Comment (ewoodev) [inline apps/examples/lcd_test/Kconfig:1]: The commit list of this PR appear to be not good..
We usually merge the PR using rebase option.
So Please update base commit of this pr branch into master branch latest commit. 

```
st7789lcd with spi
modified files with conflicts
...
Merge branch 'Samsung:master' into spi-lcd -----> do not create merge commit
...
Merge branch 'spi-lcd' of github.com:anjana348/TizenRT into spi-lcd -----> do not create merge commit
...
```

Comment (sunghan-chang) [reply apps/examples/lcd_test/Kconfig:1]: @anjana348 When you update your branch with latest public, you should use git rebase instead of git merge.
When you use git merge, it creates "Merge branch ...... into spi-lcd" commit. Actually no need in public.
Please rebase yours and then I can merge this.

File: os/drivers/lcd/st7789.c [hunk 37]
```diff
@@ -0,0 +1,901 @@
...
+ *   Sleep or wake up the driver.
+ *
+ ****************************************************************************/
+
+static void st7789_sleep(FAR struct st7789_dev_s *dev, bool sleep)
+{
+	lcd_lock(dev->spi);
+	st7789_sendcmd(dev, sleep ? ST7789_SLPIN : ST7789_SLPOUT);
+	lcd_unlock(dev->spi);
+	up_mdelay(120);
```
Comment (sunghan-chang) [inline os/drivers/lcd/st7789.c:427]: Is this required? If yes, do we need `up_mdelay` instead of `sleep`? Because `up_mdelay` is busy waiting, it's not good at performance.
